File: Construct_Binary_Tree_from_Postorder_and_inorder.java
import java.util.HashMap;

public class Construct_Binary_Tree_from_Postorder_and_inorder {
	//Approach: I use hashmap to store the indices of the nodes. Then I iterate through the inorder array, thereby getting the root from postorder traversal. Here for postorder traversal, 
	//am iterating in decreasing order to get the root.
	//2. Then , Divide the left and root children index based on the root index and recursively construct the tree. 
	HashMap<Integer, Integer> map;
    int index;
    public TreeNode buildTree(int[] inorder, int[] postorder) {
        map = new HashMap<>(); index = postorder.length-1;
        for(int i=0;i<inorder.length;i++)
        {
            map.put(inorder[i], i);
        }
        return helper(postorder, inorder, 0, inorder.length-1);
    }
    private TreeNode helper(int[] postorder, int[] inorder, int start, int end)
    {
        if(start > end || index < 0) return null;
        int idx = map.get(postorder[index]);
        TreeNode root = new TreeNode (inorder[idx]);
        index--;
        root.right = helper(postorder, inorder, idx+1, end);
        root.left = helper(postorder, inorder, start, idx-1);
        
        return root;
    }
}

//Time Complexity : O(n) where n is length of the inorder array
//Space Complexity : O(h) where h is the height of the tree. Since the recursive stack might hold elements equal to the height of the tree at worst case.
//Did this code successfully run on Leetcode : Yes
//Any problem you faced while coding this :

File: Construct_Binary_Tree_From_Postorder_and_Inorder.cpp
/*
  Construct_Tree_From_Preorder_and_Inorder.cpp
 
  Approach: Postorder gives us the root of subtree at each level in reverse order.
  Inorder gives us the left and right subtree of given root. Use hashmap to store the each node and corresponding index value from inorder array.
   
  TC: O(N), N = Number of nodes in tree
  SC: O(h + N), height of tree (maximum stack space), N = Number of nodes in tree;
*/
#include <iostream>
#include <vector>
#include <unordered_map>

using namespace std;

//Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    unordered_map<int,int> tracker;
    int id;
    TreeNode* buildTree(vector<int>& inorder, vector<int>& postorder) {
        if(inorder.size()==0)
            return NULL;
        for(int i=0;i<inorder.size();i++) tracker[inorder[i]] = i;
        id = postorder.size()-1;
        return constructTree(postorder, 0, postorder.size());
    }
   
    TreeNode* constructTree(vector<int> postorder, int start, int end){
        if(start >= end)
            return NULL;
        TreeNode* node = new TreeNode(postorder[id--]);
        int index = tracker[node->val];
        node->right = constructTree(postorder, index+1, end);
        node->left = constructTree(postorder, start, index);
        return node;
    }
     
    void inorderTraversal(TreeNode* node){
            if(node!=NULL){
                inorderTraversal(node->left);
                cout<<node->val<<"\t";
                inorderTraversal(node->right);
            }
    }
         
};

int main(int argc, const char * argv[]) {
    vector<int> postorder = {9,15,7,20,1};
    vector<int> inorder = {9,1,15,20,7};
    Solution s;
    TreeNode* root= s.buildTree(inorder, postorder);
    s.inorderTraversal(root);
    return 0;
}
